plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1006,98 +1006,106 @@ do it quite yet, even though it worked (the Plutus Tx part wasn't implemented).
     type CostingPart uni DefaultFun = BuiltinCostModel
+
+    data BuiltinVersion DefaultFun = DefaultFunV1 | DefaultFunV2
+
     -- Integers
     toBuiltinMeaning
         :: forall val. HasMeaningIn uni val
-        => DefaultFun -> BuiltinMeaning val BuiltinCostModel
-    toBuiltinMeaning AddInteger =
+        => BuiltinVersion DefaultFun -> DefaultFun -> BuiltinMeaning val BuiltinCostModel
+    toBuiltinMeaning _ver AddInteger =
         makeBuiltinMeaning
             ((+) @Integer)
             (runCostingFunTwoArguments . paramAddInteger)
-    toBuiltinMeaning SubtractInteger =
+    toBuiltinMeaning _ver SubtractInteger =
         makeBuiltinMeaning
             ((-) @Integer)
             (runCostingFunTwoArguments . paramSubtractInteger)
-    toBuiltinMeaning MultiplyInteger =
+    toBuiltinMeaning _ver MultiplyInteger =
         makeBuiltinMeaning
             ((*) @Integer)
             (runCostingFunTwoArguments . paramMultiplyInteger)
-    toBuiltinMeaning DivideInteger =
+    toBuiltinMeaning _ver DivideInteger =
         makeBuiltinMeaning
             (nonZeroArg div)
             (runCostingFunTwoArguments . paramDivideInteger)
-    toBuiltinMeaning QuotientInteger =
+    toBuiltinMeaning _ver QuotientInteger =
         makeBuiltinMeaning
             (nonZeroArg quot)
             (runCostingFunTwoArguments . paramQuotientInteger)
-    toBuiltinMeaning RemainderInteger =
+    toBuiltinMeaning _ver RemainderInteger =
         makeBuiltinMeaning
             (nonZeroArg rem)
             (runCostingFunTwoArguments . paramRemainderInteger)
-    toBuiltinMeaning ModInteger =
+    toBuiltinMeaning _ver ModInteger =
         makeBuiltinMeaning
             (nonZeroArg mod)
             (runCostingFunTwoArguments . paramModInteger)
-    toBuiltinMeaning EqualsInteger =
+    toBuiltinMeaning _ver EqualsInteger =
         makeBuiltinMeaning
             ((==) @Integer)
             (runCostingFunTwoArguments . paramEqualsInteger)
-    toBuiltinMeaning LessThanInteger =
+    toBuiltinMeaning _ver LessThanInteger =
         makeBuiltinMeaning
             ((<) @Integer)
             (runCostingFunTwoArguments . paramLessThanInteger)
-    toBuiltinMeaning LessThanEqualsInteger =
+    toBuiltinMeaning _ver LessThanEqualsInteger =
         makeBuiltinMeaning
             ((<=) @Integer)
             (runCostingFunTwoArguments . paramLessThanEqualsInteger)
     -- Bytestrings
-    toBuiltinMeaning AppendByteString =
+    toBuiltinMeaning _ver AppendByteString =
         makeBuiltinMeaning
             BS.append
             (runCostingFunTwoArguments . paramAppendByteString)
-    toBuiltinMeaning ConsByteString =
-        makeBuiltinMeaning
-            (\n xs -> BS.cons (fromIntegral @Integer n) xs)
-            (runCostingFunTwoArguments . paramConsByteString)
-    toBuiltinMeaning SliceByteString =
+    toBuiltinMeaning ver ConsByteString =
+        -- See Note [Versioned builtins]
+        case ver of
```

Comment:
good idea,  i can perhaps do this.

Reply:
Actually I can't do it with a `let impl  =` because the implementations have different type signatures. Instead I opted to share the costingfun to make it more obvious that they are the same.